src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,119 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int ELEMENT_NOT_FOUND = -1;
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        provideCapacity(size + 1);
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkAddID(index);
+        provideCapacity(size + 1);
+        System.arraycopy(elements, index, elements,
+                index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkID(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkID(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkID(index);
+        T detail = elements[index];
+        System.arraycopy(elements, index
+                + 1, elements, index, --size - index);
+        return detail;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = indexOf(element);
+        if (index != ELEMENT_NOT_FOUND) {
+            return remove(index);
+        }
+        throw new NoSuchElementException("Detail not found: " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkID(int index) {
```

Comment:
Please change ID to index everywhere. ID is common abbreviation for "identificator" and this causes confusion during code reading
```suggestion
    private void checkIndex(int index) {
```

Reply:
Hi @olekskov, I added a new code, Please check it. I hope that I did a good job now.